### PR TITLE
chore(CPlusPlus): add rod cutting problem

### PR DESCRIPTION
**Pull request by dev24il**
**PR Checklist:**
- [x] My submission is formatted according to the guidelines in the [contributing guide](https://github.com/MakeContributions/DSA#contribution-guidelines)
- [x] My addition is on refer on the language README.md file
- [x] My addition does not have a spelling problem
- [x] My submission has a proper and user-friendly description of the algorithm
- [x] My submission has the time complexity of the algorithm
- [x] My submission has sample input-output of the program (NOT FOR PYTHON)

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] New algorithm
- [ ] Optimization in previous algorithms
- [ ] Code style update
- [ ] Refactor
- [ ] Documentation
- [ ] Other, please describe:

## Rod Cutting 
- [Problem Statement](#problem-statement-for-rod-cutting)
    - [Examples](#rod-cutting-examples)
- [Optimal Approach](#rod-cutting-optimal-approach)
- [Complexity](#rod-cutting-complexity)
### Problem Statement for Rod Cutting
Given a rod of length n and a list of rod prices of length i, where 1 <= i <= n, find the optimal way to cut the rod into smaller rods to maximize profit.

image

#### Rod Cutting Examples

    1. prices[] = [1, 5, 8, 9, 10, 17, 17, 20]  
       rod length = 4  
       Best: Cut the rod into two pieces of length 2 each to gain revenue of 5 + 5 = 10
       
    2. prices[] = [1, 5, 8, 9, 10, 17, 17, 20]
       rod length = 8
       Best: Cut the rod into two pieces of length 2 and 6 to gain revenue of 5 + 17 = 22
       
    3. prices[] = [3, 5, 8, 9, 10, 17, 17, 20]
       rod length = 8
       Best: Cut the rod into eight pieces of length 1 to gain revenue of 8 * 3 = 24
       
#### Rod Cutting Optimal Approach

    We will solve this problem in a bottom-up manner. (iteratively)  
    In the bottom-up approach, we solve smaller subproblems first, then move on to larger subproblems.  
    The following bottom-up approach computes dp[i], which stores maximum profit achieved from the rod of length i from 1 to len. 
    It uses the value of smaller values i already computed.
    
image
    

**Other information:**
#### Rod Cutting Complexity
```
Time Complexity : O(n^2;)  
Space Complexity : O(n)  
```

### CODE REVIEW

**Comment by dev24il** [comment]
@Ashborn-SM please review my pr

File: algorithms/CPlusPlus/Dynamic-Programming/rod-cutting.cpp [hunk 1]
```diff
@@ -0,0 +1,74 @@
+// Rod Cutting Problem
+// Given a rod of length n and a list of rod prices of length i, where 1 <= i <= n, find the optimal way to cut the rod into smaller rods to maximize profit.
+
+// Rod Cutting Optimal Approach
+//     We will solve this problem in a bottom-up manner. (iteratively)  
+//     In the bottom-up approach, we solve smaller subproblems first, then move on to larger subproblems.  
+//     The following bottom-up approach computes dp[i], which stores maximum profit achieved from the rod of length i from 1 to len. 
+//     It uses the value of smaller values i already computed.
+
+
```
Comment (mohitchakraverty) [inline algorithms/CPlusPlus/Dynamic-Programming/rod-cutting.cpp:10]: Please also mention the time and space complexities of the program.

**Review by mohitchakraverty — CHANGES_REQUESTED** [review]

**Comment by dev24il** [comment]
@mohitchakraverty please review this pr

**Review by mohitchakraverty — APPROVED** [review]
Great work!

**Pull request merged**